keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1503 @@
+to the versioned type and then immediately performs an "unsafe" conversion to
+the internal type when possible. All subsequent request processing uses the
+internal type as the "hub" type. Validation is written for the internal
+type and so are per-resource strategies.
+
+With this change, the internal type will no longer be responsible for
+validation.
+
+If we were to convert from the internal back to the version of the API request
+for validation, we would introduce one additional conversion. If we make this
+an "unsafe" conversion, than is will be low cost for the vast majority of requests.
+
+We will benchmark this approach and plan to use it for alpha.
+
+Long term, we could do better:
+
+Since the internal type will no longer be used for validation, it becomes a lot
+less important. It is still important to have a hub type. But why not pick one
+of the versioned types to be the hub type? The vast majority of APIs only have
+one version anyway. The obvious candidate version to choose for the hub version
+would the perferred storage version.
+
+Switching to a versioned type for the hub type would have a few implications:
+
+- We would eliminate the need for internal versions.
+- We would reduce the need for "unsafe" conversions (a small security win) since
+  most APIs only have a single v1 version and so the entire request handling process
+  would simply use that version.
+- We would introduce more conversion when API request version differs from the
+  hub version. But beta APIs are off-by-default and we expect a lot less mixed
+  version API usage than in the past.
+
+This "hub version" change feels like something that could be made somewhat
+independant of this KEP.
+
+
+##### Mitigation: Avoid Conversion to Internal Type
+
+Requests are received as the versioned type, so it should be feasible to avoid
+extra conversions for resources that have no need of handwritten validations.
+
+##### Mitigation: Avoid Unstructured Conversion
+
+SMD's [value](https://github.com/kubernetes-sigs/structured-merge-diff/tree/master/value)
+wrappers have already been benchmarked and shown to have low overhead.
+
+##### Mitigation: Use Value Validations Where Possible
+
+Most validations will be possible to evaluate with simple value validations
+in native OpenAPI (required checks, enum value checks, etc). The rest are
+CEL expressions which are expected to account for 30% of all validation rule.
+
+CEL evaluation is relatively efficient we will gather detailed benchmarks baselined against native types to demonstrate this.
+
+##### Mitigation: Go fast paths for hot validation code paths
+
+If CEL becomes a bottleneck, we may explore possibilities to run Go rather than
+interpreted CEL behind the scenes for certain hot code paths:
+  - Generate Go code from CEL expressions.
+  - Hand write validation code for select validations and call those
+    directly (or from CEL).
+
+## Design Details
```

Comment:
Are you going to validate the yaml/json (or unstructured), or are you going to validate against deserialized object? There is definitely a difference between the json and the struct in which it gets deserialized. If you do the latter, i think it's fine, but you might have discrepancies between CRD and built-ins which might be hard to resolve for clients. If you do the former, you're almost certainly going to break some validation in various places. Either way, this needs to be specifically called out in the doc. Sorry if I missed it in the doc.

Reply:
> There is definitely a difference between the json and the struct in which it gets deserialized. 

While I think there would be certain beauty in performing all defaulting/validation on unstructured the same as CRDs but would result in API breaking changes that are hard to predict exhaustively.

For example, consider the field:
```
// +minimum: 1
// +optional
MyField int `json:"myField,omitempty"
```

Validating an object missing this field against `unstructured` would be fine, since `minimum` would only apply if `MyField` was specified.

However, if we deserialize first, `minimum` would fail if this field was not specified. Of course we could fix this by making it a pointer. But we have cases where non-pointer optionals exist in our API, and using pointer also has the drawback of not being able to distinguish `null` from `unset` for the rare `nullable` type.

Perhaps to transition over time to a simplified `null/empty/unset` trichotomy we could gradually:

1. Ensure all optional fields are pointers. Null will always mean UNSET.
2. Ensure all nullable optional fields are pointer-pinters or some special type of struct. If `NullableField == nil`, it is unset, but if `*NullableField == nil`, it is set to null.
3. Annotate defaults to all cases where we wanted `unset` to default to `empty`

If we could ensure these rules are followed in the API we wouldn't have to worry about Go's empty weirdness causing problems anymore, and also not break any API for users due to step 3

TLDR: For practical reasons, deserialized type. But if people care enough there is hope for a future where empty/unset/null are distinguishable.

Reply:
So for defaulting, we've treated optional "omitempty" fields with a zero-value to be "undefined".

As a side note, I've noted that you don't mention `null` anywhere in the doc, do you want to deal with that?

I suspect a solution could also be to craft a specific set of annotations that would be forbidden for omitempty fields?

Reply:
> I suspect a solution could also be to craft a specific set of annotations that would be forbidden for omitempty fields?

Not sure this would work for the case of `nullable` pointer fields to distinguish between `set to null` and `unset`.

Also relevant: https://github.com/kubernetes/kubernetes/issues/120226

CEL `has` macro does not consider `empty` to mean unset. But when we convert native types to unstructured, we have no choice to interpret `empty` to mean unset. We also have no choice but to deserialize before validating due to all defaults code taking place on the typed object. This hurdle would be removed if we switch all defaulting to declarative, I think?

Reply:
added a section about this to the doc with a plan for how to address it